### DISCUSSION TITLE
z.coerce.number() defaults empty strings to 0

### DISCUSSION THREAD

**Discussion by srowe0091** [colinhacks/zod#2814] [Q&A] [answered]
When working with react-hook-form and using controlled inputs, we need to default values to empty strings or else there will be errors of going from uncontrolled to controlled.

So using z.number() does not work because every input returns a string even if the input type is set number, it still returns a string in the event.

Switching to `z.coerce.number()` works fine with the inputs not complaining about types, but when the validation runs, empty strings are getting converted to a `0`. The state of the form shows an empty string. This causes a weird behavior where if the field is required, validation passes because the schema sees a 0 which gives a false positive.

**Comment by colinhacks** [comment]
Zod just uses the `Number` function from JavaScript to do coercion, so JavaScript is responsible for any weirdness.

What do you want `z.coerce.number().parse("")` to return?

**Comment by srowe0091** [comment]
Thanks for the reply @colinhacks, I was able to get around this by creating an actual `NumberInput` component. I have a feeling that Zod is built to in a way to make us devs write components in the right way, as opposed to just work out of the box for Inputs.

But at the time, I was hoping for it to just return the empty string as opposed to Zod, because I dont know if the zero came from the user or from Zod. So the validation of having a minimum 0 to X would always pass even though there was no value inside of the field and the user never touched the field.

**Comment by scotttrinh** [comment]
> Thanks for the reply @colinhacks, I was able to get around this by creating an actual NumberInput component. I have a feeling that Zod is built to in a way to make us devs write components in the right way, as opposed to just work out of the box for Inputs.

If I may add some color to this: Zod is built to represent the TypeScript type system at runtime. I doesn't know anything about React or HTML Forms or anything else, it's main job is to parse some value of type `unknown` into an known type providing some useful transformations on valid input data while giving you some tools to use the parsed type in the rest of your code.

The fact that HTML Form controls, in general, are very loosely/stringily typed is kind of a well-known footgun and the source for a lot of what people make fun of JavaScript about 😂 .

Having said that, I think @colinhacks 's question gets to the heart of the matter: Zod maps some input domain to a well-defined output domain, so if the output domain is `number`, what would you expect `""` to map to? Some valid values that I can think of are `0`, `NaN`, `-1`, `-Infinity`, `Infinity`, or perhaps it should refuse to coerce. Of all of those options, `0` is the closest to being the natural fit since that's what JavaScript's `Number` constructor casts `""` as.

---

In my view, form libraries should be doing the heavy lifting for you here: setting "empty" values to `null` and doing whatever transformations to the values are required to make using parsing/validation libraries like Zod give sensible behavior based on the HTML Form use-case.

**Comment by srowe0091** [comment]
@scotttrinh Totally understand. It's been a pretty well known problem and massive disconnect when dealing with types and values coming back from input fields. When setting fields to null/undefined and then add a value (if its controlled) then you get the error that a field has been changed from uncontrolled to controlled. so the empty string was the only logical thing to add as the default value and to also keep the input field empty (0 wouldn't work here because the field should be empty). so thats why we're here lol.

but i totally understand your point and how Zod was developed. I just have to be more explicit about when dealing with the types of data i define in the schema and what components will be used for that type.

This issue can be closed since this isn't an issue for Zod to fix anyway.

For anyone curious what i did, I just created a `NumberInput` component that overrides the onChange and then casts the string in the `event.target.value` into a number `Number(event.target.value)`. and if its an empty string `""` then i return an empty string, which will then get processed in the validation saying its the wrong type, but i just override that message saying `Invalid input`

**Comment by alainfonhof** [comment]
We had a similar issue and settled on `z.coerce.number().positive()`

**Comment by daniel-johnsson** [comment]
Several UI libraries use `""` for number inputs, would be great if coerce would parse empty string to undefined to not trigger the invalid_type_error and rather required_error if not set to optional. 

My current workaround to the problem is to extend the `ZodNumber` class :

``` typescript
import {
  ParseInput,
  ParseReturnType,
  ProcessedCreateParams,
  RawCreateParams,
  ZodErrorMap,
  ZodFirstPartyTypeKind,
  ZodNumber,
} from "zod";

// Directly copied from zod/src/types.ts
function processCreateParams(params: RawCreateParams): ProcessedCreateParams {
  if (!params) return {};
  const { errorMap, invalid_type_error, required_error, description } = params;
  if (errorMap && (invalid_type_error || required_error)) {
    throw new Error(
      `Can't use "invalid_type_error" or "required_error" in conjunction with custom error map.`
    );
  }
  if (errorMap) return { errorMap: errorMap, description };
  const customMap: ZodErrorMap = (iss, ctx) => {
    if (iss.code !== "invalid_type") return { message: ctx.defaultError };
    if (typeof ctx.data === "undefined") {
      return { message: required_error ?? ctx.defaultError };
    }
    return { message: invalid_type_error ?? ctx.defaultError };
  };
  return { errorMap: customMap, description };
}

export class CustomZodNumber extends ZodNumber {
  _parse(input: ParseInput): ParseReturnType<number> {
    // Alot of input ui libraries will send empty strings instead of undefined
    // This is a workaround for not triggering invalid_type_error and rather required_error if not optional
    input.data = input.data === "" ? undefined : input.data;
    return super._parse(input);
  }
  static create = (
    params?: RawCreateParams & { coerce?: boolean }
  ): CustomZodNumber => {
    return new CustomZodNumber({
      checks: [],
      typeName: ZodFirstPartyTypeKind.ZodNumber,
      coerce: params?.coerce || false,
      ...processCreateParams(params),
    });
  };
}

export const number = CustomZodNumber.create;
```

Would love if `processCreateParams` where a exported function as well so I did not have to copy it.

**Comment by jacknevitt** [comment]
Is there a way to validate that the is not an empty string before coercing?
My best guess is to use `.pipe`?
```ts
z.number().or(z.string().nonempty())
  .pipe(z.coerce.number())

// Or only in the context of form inputs
z.string().nonempty()
  .pipe(z.coerce.number())
```
And for optional fields
```ts
z.literal("").transform(() => undefined)
  .or(z.coerce.number())
  .optional()
```

**Comment by colinhacks** [comment]
@jacknevitt That would work but `z.preprocess()` would be a little more explicit.

**Comment by jacknevitt** [comment]
Oh yeah, of course. I had forgotten about that one. Thanks!

**Comment by maximeburri** [comment]
For anyone interested, I found this solution (but without using `coerce`):

```js
const ZodStringNumberOrNull = z
  .string()
  .transform((value) => (value === '' ? null : value))
  .nullable()
  .refine((value) => value === null || !isNaN(Number(value)), {
    message: 'Invalid number',
  })
  .transform((value) => (value === null ? null : Number(value)));
```

This returns `number | null` and handles cases:
- `''` parses `null`
- `'100'` parses `100`
- `'100a'` will fail when parsing with `Invalid number`

**Reply by Korat-Dishant** [reply]
This would fail if the input is number. would give error like 'Expected string, received number'

I am use react hook form with zod. After fetching data from database I pass it to default values. So this default value would be of type number instead of number wrapped in string. If user makes some changes in HTML form field with number then it there won't be any problem since HTML input type=number, converts input to string and your code would work. but say if user  dont make changes to the field and try to submit the form then your schema would receive the number and would throw error.

Any idea how to get around this? Already tried preprocess but it is inferring types as 'unknown' answers useWatch and watch also infers 'unknown'

**Reply by marcomuser** [reply]
If you use a preprocess step you can work around this easily:
```ts
z.preprocess((v) => {
    if (typeof v !== "string") {
      return v;
    }
    if (v.trim() === "") {
      return undefined;
    }
    return Number(v);
  }, z.number());
```

If you have a number it would just pass that number to your schema. If your user makes changes, the preprocessor will receive a string and try to convert that to a number before passing it to your schema.

Preprocessor example taken from: https://github.com/marcomuser/conformal/blob/84c74334b937213104cd5577fab9609791a5bee2/src/zod/schemas.ts#L15C1-L25C2

**Reply by Korat-Dishant** [reply]
Thanks for the suggestion.  I am using preprocessor in my code but it have some problems

```export const number = (options?: {
  required?: boolean;
  min?: number;
  max?: number;
  minMessage?: string;
  maxMessage?: string;
  requiredMessage?: string;
}) => {
  const {
    required = false,
    max,
    min,
    maxMessage = "Number too big",
    minMessage = "Number too small",
    requiredMessage = "This field is required",
  } = options ?? {};

  let baseSchema = z.number({
    invalid_type_error: "Must be a valid number",
    required_error: requiredMessage,
  });

  if (!isEmpty(min)) {
    baseSchema.min(min, { message: minMessage });
  }
  if (!isEmpty(max)) {
    baseSchema.max(max, { message: maxMessage });
  }

  const schema = required ? baseSchema : baseSchema.optional();

  return z.preprocess((val) => {
    if (isEmpty(val)) {
      return undefined;
    }
    if (typeof val === "number") {
      return val;
    }
    if (typeof val === "string") {
      const num = Number(val);
      return isNaN(num) ? val : num;
    }
    return val;
  }, schema);
};
```




But the problem with using preprocess is that it infers the type as unknown. 

```const testSchema = z.object({
  favNumber: number({ required: false, max: 100, min: 0 }),
});
type formType = z.infer<typeof testSchema>;
type formOutputType = z.output<typeof testSchema>;
type formInputType = z.input<typeof testSchema>;
const {
  register,
  handleSubmit,
  control,
  watch,
  setValue,
  formState: { errors: rhfErrors, defaultValues },
} = useForm({
  resolver: zodResolver(testSchema),
  defaultValues: { favNumber: "", stringNumber: 66 },
});

const watchedFavNumber = watch("favNumber");
```



IF you hover over formInputType you would see 
<img width="563" height="323" alt="Screenshot 2025-09-24 at 3 42 13 PM" src="https://github.com/user-attachments/assets/6a169325-e734-47b6-8204-8be20638fa86" />
```
favNumber?: unknown;
```
and  in formType  = z.infer<typeof testSchema>;
```
    favNumber?: number | undefined;
```


This cause problem with react hook form. as const watchedFavNumber = watch("favNumber"); would give you 
```
const watchedFavNumber: unknown
```

**Reply by marcomuser** [reply]
Very difficult to get the full picture of what you are doing from the screenshots. If you try something simple like this, favNumber will be correctly inferred as a number:
```tsx
const formSchema = z.object({
  favNumber: z.preprocess((v) => {
    if (typeof v !== 'string') {
      return v;
    }
    if (v.trim() === '') {
      return undefined;
    }
    return Number(v);
  }, z.number()),
});
type FormFields = z.infer<typeof formSchema>;

export default function App() {
  const { control, handleSubmit } = useForm<FormFields>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    resolver: zodResolver(formSchema),
    defaultValues: {
      favNumber: 2
    },
  });
```

**Reply by Korat-Dishant** [reply]
I tried your suggested code but resolver is generating the following ts error 
```
Type 'Resolver<{ favNumber?: unknown; }, any, { favNumber?: number; }>' is not assignable to type 'Resolver<{ favNumber?: number; }, any, { favNumber?: number; }>'.
  Types of parameters 'options' and 'options' are incompatible.
    Type 'ResolverOptions<{ favNumber?: number; }>' is not assignable to type 'ResolverOptions<{ favNumber?: unknown; }>'.
      Type '{ favNumber?: unknown; }' is not assignable to type '{ favNumber?: number; }
```

Sorry for the code provided earlier It was directly provided from my codebase with few modifications. 
You can see the problem I am talking about in this dev environment.

https://stackblitz.com/edit/react-ts-lytwhvvn?devToolsHeight=33&file=App.tsx




Here is how I want my number input field to behave
-> if number is passed as string and able to convert it to number raise no error
-> if random string is passed and unable to convert to number raise validation error
-> if empty string is passed "" then consider it as undefined, not 0
-> if number is passed as number raise no error
-> Provide valid types (avoid force type casting if possible), I want the field to take 'number' | 'string' | undefined | null as input and provide 'number' | undefined as output

**Reply by marcomuser** [reply]
Thanks for the Stackblitz. Looks like a bug in the new major version of @hookform/resolvers to me. If I downgrade to the latest v4 version it works as expected.

**Reply by tevinch** [reply]
For the input/output typing follow-up in this thread, Zod 4 supports annotating the preprocessor parameter directly. Keep the output schema optional *inside* preprocessing:

```ts
import { z } from 'zod';

const optionalNumber = z.preprocess(
  (value: string | number | null | undefined) => {
    if (value === null || value === undefined) return undefined;
    if (typeof value !== 'string') return value;
    const text = value.trim();
    return text === '' ? undefined : Number(text);
  },
  z.number().optional(),
);

type Input = z.input<typeof optionalNumber>; // string | number | null | undefined
type Output = z.output<typeof optionalNumber>; // number | undefined

const schema = z.object({ favNumber: optionalNumber });

// Inside the React component:
const form = useForm<z.input<typeof schema>, unknown, z.output<typeof schema>>({
  resolver: zodResolver(schema),
  defaultValues: { favNumber: '' },
});
```

Import `useForm` from `react-hook-form` and `zodResolver` from `@hookform/resolvers/zod`. For a controlled input, use `field.value ?? ''` and pass its change event to `field.onChange`. This keeps numeric `0` visible and lets the resolver own conversion. `watch` has the input type; a successful `handleSubmit` callback has the output type.

For a required value, `optionalNumber.pipe(z.number())` rejects empty values while allowing zero. This policy treats null as empty and retains JavaScript `Number` syntax for nonempty strings; it does not parse locale formatting or preserve exact decimal precision.

I also put together a free [MIT example with runtime/type checks and a controlled form](https://github.com/tevinch/data-shape-kit/tree/main/examples/number-input), tested with Zod 4.6.1, resolvers 5.9.1 and React Hook Form 7.87.0. It builds on the preprocessing approach already discussed here. The relevant official references are [Zod input annotations](https://zod.dev/api#preprocess) and [resolver input/output generics](https://github.com/react-hook-form/resolvers#typescript).

**Comment by LouisLecouturier** [comment]
Using @maximeburri approach, I made a function that wraps the `z.number()` type you'd like :

```ts
import { z, ZodTypeAny } from 'zod';

export const zodInputStringPipe = (zodPipe: ZodTypeAny) =>
  z
    .string()
    .transform((value) => (value === '' ? null : value))
    .nullable()
    .refine((value) => value === null || !isNaN(Number(value)), {
      message: 'Nombre Invalide',
    })
    .transform((value) => (value === null ? 0 : Number(value)))
    .pipe(zodPipe);
```

In use :

```ts
const schema = zodInputStringPipe(z.number().positive('Le nombre doit être supérieur à 0'));
```

May be handier to use in your schemas

**Reply by tyteen4a03** [reply]
Do you have the TypeScript type for this?

**Comment by edersonlucas** [comment, marked as answer]
> Using @maximeburri approach, I made a function that wraps the `z.number()` type you'd like :
> 
> ```ts
> import { z, ZodTypeAny } from 'zod';
> 
> export const zodInputStringPipe = (zodPipe: ZodTypeAny) =>
>   z
>     .string()
>     .transform((value) => (value === '' ? null : value))
>     .nullable()
>     .refine((value) => value === null || !isNaN(Number(value)), {
>       message: 'Nombre Invalide',
>     })
>     .transform((value) => (value === null ? 0 : Number(value)))
>     .pipe(zodPipe);
> ```
> 
> In use :
> 
> ```ts
> const schema = zodInputStringPipe(z.number().positive('Le nombre doit être supérieur à 0'));
> ```
> 
> May be handier to use in your schemas

Thank you, that's exactly what I was looking for.

**Comment by srowe0091** [comment]
> @srowe0091, Has your question been addressed? If so, I'd like to close this issue.

@JacobWeisenburger Yes its been addressed, thank you!

**Reply by JacobWeisenburger** [reply]
Please select the answer that solved your issue.

**Comment by Lukem121** [comment]
The following example shows how you can use Zod with React Hook form with the default values being empty strings. 

In the example:
dogs is a required number.
cats is a optional number.

```typescript
const FormSchema = z
  .object({
    dogs: z.union([
      z.coerce
        .number({
          message: "must be a number",
        })
        .int({
          message: "must be a whole number",
        })
        .positive({
          message: "must be positive",
        }),
      z.literal("").refine(() => false, {
        message: "is required",
      }),
    ]),
    cats: z
      .union([
        z.coerce
          .number({
            message: "must be a number",
          })
          .int({
            message: "must be a whole number",
          })
          .positive({
            message: "must be positive",
          }),
        z.literal("").refine(() => false, {
        message: "is required",
      }),
      ])
      .optional(),
  })
  
  
  const form = useForm<z.infer<typeof FormSchema>>({
    resolver: zodResolver(FormSchema),
    defaultValues: {
      dogs: "",
      cats: "",
    },
  });
```

**Reply by scottymccormick** [reply]
@Lukem121 Your validation for `dogs` here still allows an empty string that is coerced to 0.

**Reply by Lukem121** [reply]
I've just tested this and when I parse an empty string I get an error. It does not seem to be coercing the empty string into a zero. Please let me know your findings.

```
const parsed = FormSchema.safeParse({
    dogs: "",
    cats: "",
});

if (parsed.error) {
    console.log(parsed.error);
}
```

The console.log shows:
```
ZodError: [
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": false,
    "exact": false,
    "message": "must be positive",
    "path": [
      "dogs"
    ]
  },
  {
    "code": "too_small",
    "minimum": 0,
    "type": "number",
    "inclusive": false,
    "exact": false,
    "message": "must be positive",
    "path": [
      "cats"
    ]
  }
]
```

**Reply by scottymccormick** [reply]
@Lukem121 Ah, my bad, was testing without the `.positive()` constraint on my end. I'm trying to achieve this while allowing zero and negative numbers.

**Reply by Lukem121** [reply]
Not sure if this is the best way but this is how I have achived it while allowing zero and negative numbers.

```
const FormSchema = z.object({
    dogs: z.preprocess(
      (val) => {
        if (val === "") {
          return null; // Treat empty string as null 
        }
        return val;
      },
      z.union([
        z.null().refine(() => false, {
          message: "is required",
        }),
        z.coerce
          .number({
            message: "must be a number",
          })
          .int({
            message: "must be a whole number",
          }),
      ]),
    ),
  });
```

**Comment by juarezs** [comment]
this works and keeps the field typed as `number` only (instead of string | number)

```
    z
    .number()
    .transform((value) => (!value ? null : value))
    .nullish()
    .or(z.coerce.number({message: "must be a number."}))
    .nullish()
```

**Comment by jhelmink** [comment]
I wanted something similar but it should support a nullable number input

```
export const zodNullableNumberInput = () =>
    z.coerce // Let zod try to coerce a numeric input to a string
        .string() // Input starts as a string
        .transform((value) => (value === "" ? null : value)) // Empty string treated as null
        .refine((value) => !isNaN(Number(value)), { message: "Must be a number." }) // Otherwise test if the string is a number
        .transform((value) => (value === null ? null : Number(value))) // If it's not null and is a number return it as a number
        .nullable(); // This can be a number or null
```

**Comment by satnaing** [comment]
Although I found the marked answer and others helpful, I had different requirements.

Mine is a `number` input and I want the following rules & requirements.
- must be `type="number"` input
- must allow `0`, `negative` and `positive` values
- should be able to set empty string (`""`) as the default
- should be able to set numbers as the default
- must not display "controlled/uncontrolled" error when `onChange`

And here's my solution

```ts
totalPercentage: z.union([
  z.coerce
    .string()
    .min(1, { message: "Total Percentage is required." })
    .transform((v) => Number(v)),
  z.literal("").refine(() => false, { message: "Total Percentage is required." }),
])
```

You could also make this as a reusable function.

```ts
const zodNumType = (message: string) =>
  z.union([
    z.coerce
      .string()
      .min(1, { message })
      .transform((v) => Number(v)),
    z.literal("").refine(() => false, { message }),
  ])
```

Then, use the above function like this

```ts
z.object({
  totalPercentage: zodNumType("Total Percentage is required."),
  monthlyPercentage: zodNumType("Monthly Percentage is required."),
})
```

> the inferred type will be `number | ""`

**Comment by scottwillman** [comment]
I think I found a much cleaner and simpler approach to treat empty number strings as null, throw errors, and still handle all numeric validation natively. Use `preprocess` instead of `coerce` and then (optionally) chain your secondary validations after the second argument like so:

### Required
```javascript
const validator = z.object({
    numberInput: z
        .preprocess(
            (value) => (value === '' ? null : Number(value)), 
            z.number({ message: 'A number is required' })
                .positive({ message: 'Number must be positive' })
        )
});
```

### Optional
note that empty strings now return `undefined`
```javascript
const validator = z.object({
    numberInput: z
        .preprocess(
            (value) => (value === '' ? undefined : Number(value)), 
            z.number({ message: 'A number is required' })
                .positive({ message: 'Number must be positive' })
                .optional()
        )
});
```

**Reply by kesitrifork** [reply]
Also me 👍 preprocess is better, I also ended up with something to parse searchParams

Where a empty url would return the default, since `url.searchParams.get('query')` will return null if query is not present.
 
```
z.preprocess((value) => {
		if(value === null) return undefined
		return value === '' ? undefined : Number(value);
	}, z.number().min(1).max(10).default(3))
```

**Comment by daveycodez** [comment]
KISS

```ts
z.preprocess(
    (val) => (!val ? undefined : Number(val)),
    z.number({
        required_error: `Field is required`,
        invalid_type_error: `Field must be a number`
    })
)
```

**Reply by marc-niclas** [reply]
> KISS
> 
> ```ts
> z.preprocess(
>     (val) => (!val ? undefined : Number(val)),
>     z.number({
>         required_error: `Field is required`,
>         invalid_type_error: `Field must be a number`
>     })
> )
> ```

Didn't work for me because of typing. `z.input` returns a type of `unknown` for this field.

The accepted answer gets pretty close, however the type of `z.infer` returns `number | undefined`, even if my pipe is a required number. The only way I could get this to work is with some Typescript hacks:
```typescript
export const zodInputStringPipe = <
  P extends
    | z.ZodNumber
    | z.ZodOptional<z.ZodNumber>
    | z.ZodEffects<z.ZodNumber, number, number>
    | z.ZodEffects<z.ZodOptional<z.ZodNumber>, number, number>,
>(
  zodPipe: P,
): ZodInputStringPipe<P> =>
  z
    .string()
    .transform((value) => (value === '' ? undefined : value))
    .optional()
    .refine((value) => value === undefined || !Number.isNaN(Number(value)), {
      message: 'Valid number is required',
    })
    .transform((value) => (value === undefined ? undefined : Number(value)))
    .pipe(zodPipe) as ZodInputStringPipe<P>;
```

With this, the `z.infer` type for `zodInputStringPipe(z.number())` returns `number`, while the `z.infer` type for `zodInputStringPipe(z.number().optional())` returns `number | undefined`.

`z.input` returns a type of `string | undefined` for this field but thats ok for my purposes.

EDIT: v4 has different types:
```typescript
type OptionalStringInput =
  | z.ZodNumber
  | z.ZodOptional<z.ZodNumber>
  | z.ZodPipe<z.ZodNumber, z.ZodTransform<number, number>>;

type ZodInputStringPipe<P extends OptionalStringInput> = z.ZodPipe<
  z.ZodPipe<z.ZodOptional<z.ZodPipe<z.ZodString, z.ZodTransform<string | undefined, string>>>>,
  P
>;

export const zodInputStringPipe = <P extends OptionalStringInput>(zodPipe: P) =>
  z
    .string()
    .transform((value) => (value === '' ? undefined : value))
    .optional()
    .refine((value) => value === undefined || !Number.isNaN(Number(value)), {
      message: 'Valid number is required',
    })
    .transform((value) => (value === undefined ? undefined : Number(value)))
    .pipe(zodPipe) as unknown as ZodInputStringPipe<P>;
```

**Comment by yulafezmesi** [comment]
how can i achieve this with using zod/v4-mini?

https://github.com/colinhacks/zod/discussions/2814#discussioncomment-7121769

**Comment by marcomuser** [comment]
I've worked on a comprehensive set of preprocessors for use of zod with html form values. The number preprocessor would treat trimmed empty strings as undefined before coercion to handle the issue mentioned.

I'm currently exploring that path as part of my conformal lib but I think if that proves useful it could also live under a z.formdata subobject directly in zod.

Take a look: https://github.com/marcomuser/conformal/tree/main/src/zod
